File: sdk/typescript/src/runtime.ts
export interface ResultEnvelope {
  protocol: "aclip/0.1";
  type: "result";
  ok: true;
  command: string;
  data: Record<string, unknown>;
}

export const AUTH_ERROR_CODES = [
  "auth_required",
  "invalid_credential",
  "expired_credential"
] as const;

export interface ErrorEnvelope {
  protocol: "aclip/0.1";
  type: "error";
  ok: false;
  command: string;
  error: {
    code: string;
    message: string;
    category?: string;
    retryable?: boolean;
    hint?: string;
  };
}

export interface ErrorEnvelopeOptions {
  category?: string;
  retryable?: boolean;
  hint?: string;
}

export function encodeJson(payload: unknown): string {
  return `${JSON.stringify(payload)}\n`;
}

export function resultEnvelope(command: string, data: unknown): ResultEnvelope {
  return {
    protocol: "aclip/0.1",
    type: "result",
    ok: true,
    command,
    data: isRecord(data) ? data : { result: data }
  };
}

export function renderSuccessOutput(data: unknown): string {
  if (data === undefined || data === null) {
    return "";
  }
  if (typeof data === "string") {
    return data.endsWith("\n") ? data : `${data}\n`;
  }
  return encodeJson(data);
}

export function errorEnvelope(
  command: string,
  code: string,
  message: string,
  options: ErrorEnvelopeOptions = {}
): ErrorEnvelope {
  return {
    protocol: "aclip/0.1",
    type: "error",
    ok: false,
    command,
    error: {
      code,
      message,
      ...(options.category ? { category: options.category } : {}),
      ...(options.retryable !== undefined ? { retryable: options.retryable } : {}),
      ...(options.hint ? { hint: options.hint } : {})
    }
  };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: sdk/python/src/aclip/runtime.py
from __future__ import annotations

import json
from typing import Any

AUTH_ERROR_CODES = {
    "auth_required",
    "invalid_credential",
    "expired_credential",
}


def encode_json(payload: Any) -> str:
    return json.dumps(payload, ensure_ascii=True, sort_keys=True)


def result_envelope(command: str, data: Any) -> dict[str, Any]:
    if not isinstance(data, dict):
        data = {"result": data}
    return {
        "protocol": "aclip/0.1",
        "type": "result",
        "ok": True,
        "command": command,
        "data": data,
    }


def render_success_output(data: Any) -> str:
    if data is None:
        return ""
    if isinstance(data, str):
        return data if data.endswith("\n") else f"{data}\n"
    return f"{encode_json(data)}\n"


def error_envelope(
    command: str,
    code: str,
    message: str,
    *,
    category: str | None = None,
    retryable: bool | None = None,
    hint: str | None = None,
) -> dict[str, Any]:
    error = {
        "code": code,
        "message": message,
    }
    if category is not None:
        error["category"] = category
    if retryable is not None:
        error["retryable"] = retryable
    if hint is not None:
        error["hint"] = hint

    return {
        "protocol": "aclip/0.1",
        "type": "error",
        "ok": False,
        "command": command,
        "error": error,
    }
